qcodes/instrument_drivers/Keysight/Infiniium.py

```diff
@@ -1,646 +1,1103 @@
+    If you need to ensure a fresh value, run dso.digitize() prior to reading
+    the measurement value.
     """
-    # note: this is not really a channel, but InstrumentChannel does everything
-    # a 'Submodule' class should do
 
-    def __init__(self, parent: Instrument, name: str, **kwargs: Any) -> None:
+    def __init__(self, parent: InstrumentBase, name: str, **kwargs: Any) -> None:
+        """
+        Add parameters to measurement subsystem. Note: This should not be initialized
+        directly, rather initialize BoundMeasurementSubsystem
+        or UnboundMeasurementSubsystem.
+        """
         super().__init__(parent, name, **kwargs)
 
-        self.add_parameter(name='source_1',
-                           label='Measurement primary source',
-                           set_cmd=partial(self._set_source, 1),
-                           get_cmd=partial(self._get_source, 1),
-                           val_mapping={i: f'CHAN{i}' for i in range(1, 5)},
-                           snapshot_value=False)
-
-        self.add_parameter(name='source_2',
-                           label='Measurement secondary source',
-                           set_cmd=partial(self._set_source, 2),
-                           get_cmd=partial(self._get_source, 2),
-                           val_mapping={i: f'CHAN{i}' for i in range(1, 5)},
-                           snapshot_value=False)
-
-        self.add_parameter(name='amplitude',
-                           label='Voltage amplitude',
-                           get_cmd=self._make_meas_cmd('VAMPlitude'),
-                           get_parser=float,
-                           unit='V',
-                           snapshot_value=False)
-
-        self.add_parameter(name='average',
-                           label='Voltage average',
-                           get_cmd=self._make_meas_cmd('VAVerage'),
-                           get_parser=float,
-                           unit='V',
-                           snapshot_value=False)
-
-        self.add_parameter(name='base',
-                           label='Statistical base',
-                           get_cmd=self._make_meas_cmd('VBASe'),
-                           get_parser=float,
-                           unit='V',
-                           snapshot_value=False)
-
-        self.add_parameter(name='frequency',
-                           label='Signal frequency',
-                           get_cmd=self._make_meas_cmd('FREQuency'),
-                           get_parser=float,
-                           unit='Hz',
-                           docstring="""
+        ###################################
+        # Voltage Parameters
+        self.add_parameter(
+            name="amplitude",
+            label="Voltage amplitude",
+            get_cmd=self._create_query("VAMP"),
+            get_parser=float,
+            unit="V",
+            snapshot_value=False,
+        )
```

Comment:
Since you are rewriting the driver you may want to consider using the new form of parameter declaration

As in this example

https://qcodes.github.io/Qcodes/examples/writing_drivers/Creating-Instrument-Drivers.html#VisaInstrument:-Simple-example

This has the benefits that tap completion works correctly for parameters and the parameters can be documented as attributes. See https://qcodes.github.io/Qcodes/api/generated/qcodes.instrument_drivers.weinschel.html#qcodes.instrument_drivers.weinschel.Weinschel_8320.Weinschel_8320.attenuation for an example

Reply:
Hey @jenshnielsen - implemented this in a8f49ad

Unfortunately this does cause a couple of new mypy errors to appear, since the results of a parameter get (which are used in a couple of places) are not defined...

For example, Infiniium.py#115 now fails the mypy check, since we query the parameter `dso.funcX.function` to figure out whether we are measuring an FFT.

Any suggestions on how to fix/correctly type annotate parameters?

Reply:
I think https://github.com/QCoDeS/Qcodes/pull/3935 should resolve that by clarifying the difference of get and set. The optional part is for set.

Reply:
There is a remaining issue of the form
```
qcodes\instrument_drivers\Keysight\Infiniium.py:109:31: error: Cannot determine type of "cache_setpoints"  [has-type]
qcodes\instrument_drivers\Keysight\Infiniium.py:190:16: error: Cannot determine type of "cache_setpoints"  [has-type]
qcodes\instrument_drivers\Keysight\Infiniium.py:200:12: error: Cannot determine type of "auto_digitize"  [has-type]
```

It looks like mypy cannot always infer the type of the attribute correctly
This can be mitigated by doing:

![image](https://user-images.githubusercontent.com/548266/154274124-07f45db6-93da-47f0-be16-6131228c3bb9.png)

Reply:
We could also ignore the error on those lines. Hopefully this bug in mypy will eventually be resolved

Reply:
@astafan8 Yes but that does not fix the issue. An alternative could be to just ignore the mypy warning

Reply:
@astafan8 `Cannot determine type of` does not mean mypys need a type annotation here but something more like mypy cannot figure out what has happened here and is giving up. Adding a type annotation does not help

Reply:
ah, that's indeed weird and unexpected.

Reply:
@astafan8 Actually it does seem like you are right 

Doing

self.cache_setpoints: Parameter = Parameter(...)

does fix the issue.

Reply:
I have created a pr here https://github.com/spauka/Qcodes/pull/18 that makes it type check against master. Note that you need to also rebase or merge with master for that to work

Reply:
Success ♥️

Thanks for the suggestions for how to solve. I'm surprised that the type annotation of `self.cache_setpoints` worked given that it's literally during the definition of the parameter as a `Parameter` 🤣